### PR TITLE
[pbh]: Add PBH OA

### PR DESCRIPTION
**Pull request by nazariig**
**PBH architecture:**
1. PBH OA
2. PBH Helper
3. PBH Container

**PBH business logic:**
1. PBH OA: receive the updates from Redis (Config DB)
2. PBH Helper/PBH Container: parse data/store in internal cache:

    - If no errors observed - store task by table key in internal cache
    - If task with the same key already exists - postpone processing and proceed with the next one
    - Set/Remove tasks are stored in separate maps
    - For each type of Config DB entity (table, rule, hash and hash field) a dedicated PBH container is used

3. PBH OA: deploy tasks:

    - Use Remove & Setup policy
    - Process create/update (SET) tasks in correct order
    - Do the same for delete (REMOVE) tasks

#### Why I did it
* Done as part of PBH development

#### How I did it
* Added PBH OA

#### How to verify it
1. Add PBH configuration
2. Run DPB

#### Which release branch to backport (provide reason below if selected)

- [X] 202106

#### Description for the changelog

* N/A

#### A picture of a cute animal (not mandatory but encouraged)
```
      .---.        .-----------
     /     \  __  /    ------
    / /     \(  )/    -----
   //////   ' \/ `   ---
  //// / // :    : ---
 // /   /  /`    '--
//          //..\\
       ====UU====UU====
           '//||\\`
             ''``
```

### CODE REVIEW

File: orchagent/aclorch.cpp
Comment (prsunny) [inline orchagent/aclorch.cpp:1418]: can you move this pbh ACLs to a separate file?

**Comment by prsunny** [comment]
@anish-n to review and sign-off

File: orchagent/aclorch.h
Comment (stepanblyschak) [inline orchagent/aclorch.h:428]: Actually this mutex is no longer needed. Few releases back orchagent spawned a thread for counter polling and this mutex was introduced. Later approach with SelectableTimer was used. So, orchagent is fully single threaded now.

Comment (nazariig) [reply orchagent/aclorch.h:428]: @stepanblyschak Thanks! Legacy code is always confusing...

File: orchagent/pbhorch.cpp
Comment (anish-n) [inline orchagent/pbhorch.cpp:904]: You are checking this in caller(deployPbhHashFieldSetupTasks) already, any value in adding this check in both places?

Comment (anish-n) [reply orchagent/pbhorch.cpp:904]: Similar comment applies for other fns in this file as well, eg: updatePbhHashField and others

File: tests/test_pbh.py
Comment (anish-n) [inline tests/test_pbh.py:249]: Do we have any tests for retry flows where some object is created before a dependency is met?

File: orchagent/pbh/pbhmgr.cpp
Comment (anish-n) [inline orchagent/pbh/pbhmgr.cpp:197]: As far as I see the purpose of this function is to decrementAllDependencyRefCountsIs and not removeDependencies, I would recommend changing it per purpose of the function.

Comment (anish-n) [reply orchagent/pbh/pbhmgr.cpp:197]: Similar comment applies to other types of Pbh objects as well

File: orchagent/pbhorch.cpp
Comment (nazariig) [reply orchagent/pbhorch.cpp:904]: @anish-n this is an additional check just in case if the method will be called from another API. For now we can remove it as you mentioned.

File: tests/test_pbh.py
Comment (nazariig) [reply tests/test_pbh.py:249]: @anish-n no, i will add.

File: orchagent/pbh/pbhmgr.cpp
Comment (nazariig) [reply orchagent/pbh/pbhmgr.cpp:197]: @anish-n what other types of PBH objects? Can you please elaborate?

As for your comment we can do:
```cpp
bool hasDependencies(const PbhContainer &obj) const;
template<typename T>
bool validateDependencies(const T &obj) const;

template<typename T>
bool incRefCount(const T &obj);
template<typename T>
bool decRefCount(const T &obj);
```

Is that what you are expecting?

File: orchagent/aclorch.cpp
Comment (nazariig) [reply orchagent/aclorch.cpp:1418]: @prsunny will do.

File: orchagent/aclorch.cpp
Comment (prsunny) [inline orchagent/aclorch.cpp:158]: @bingwang-ms , could you please take a look at the `aclorch` changes?

**Comment by nazariig** [comment]
@anish-n please review and sign-off

**Comment by nazariig** [comment]
@bingwang-ms please review and sign-off

**Review by anish-n — APPROVED** [review]
PBH-wise looks good to me. Please wait for review from Bing on ACLorch and Prince on overall Orchagent prior to merge.

**Comment by nazariig** [comment]
@bingwang-ms just a kind reminder

**Comment by anshuv-mfst** [comment]
Hi @prsunny - could you please review and approve asap, this is needed in 202106 release

File: orchagent/aclorch.cpp
Comment (bingwang-ms) [inline orchagent/aclorch.cpp:805]: Do we need to check the existance of counter before disable it?

File: orchagent/aclorch.cpp
Comment (bingwang-ms) [inline orchagent/aclorch.cpp:785]: If SAI return error for enbling counter, do we need to remove the already created counter?

File: orchagent/aclorch.cpp
Comment (bingwang-ms) [inline orchagent/aclorch.cpp:3256]: I think this interface is not covered by unit test. Could you please check?

**Comment by bingwang-ms** [comment]
> @bingwang-ms just a kind reminder

Sorry for missing the review. Most part of aclorch looks good to me. Please check the comments I left. Thanks

File: orchagent/aclorch.cpp
Comment (nazariig) [reply orchagent/aclorch.cpp:3256]: @bingwang-ms sure. I will add a UT.

File: orchagent/aclorch.cpp
Comment (nazariig) [reply orchagent/aclorch.cpp:785]: @bingwang-ms would be better to remove to save system resources

File: orchagent/aclorch.cpp
Comment (nazariig) [reply orchagent/aclorch.cpp:805]: @bingwang-ms it is not mandatory, however we can save some time on SAI API call

Comment (nazariig) [reply orchagent/aclorch.cpp:805]: @bingwang-ms would be better to remove to save system resources

File: orchagent/aclorch.cpp
Comment (nazariig) [reply orchagent/aclorch.cpp:785]: @bingwang-ms would be better to remove to save system resources

File: orchagent/aclorch.cpp
Comment (nazariig) [reply orchagent/aclorch.cpp:805]: @bingwang-ms it is not mandatory, however we can save some time on SAI API call

File: orchagent/aclorch.cpp
Comment (nazariig) [reply orchagent/aclorch.cpp:790]: @bingwang-ms done

File: orchagent/aclorch.cpp
Comment (nazariig) [reply orchagent/aclorch.cpp:816]: @bingwang-ms done

File: orchagent/aclorch.cpp
Comment (nazariig) [reply orchagent/aclorch.cpp:3269]: @bingwang-ms done

**Review by bingwang-ms — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@anish-n please approve.
@nazariig please check failures. if common lets take with @prsunny

**Comment by nazariig** [comment]
@prsunny we have got all the approvals, please review & merge

**Comment by prsunny** [comment]
@anish-n , @bingwang-ms , could you please sign-off?

**Review by bingwang-ms — APPROVED** [review]

**Review by anish-n — APPROVED** [review]

**Pull request merged**
